### PR TITLE
Fix inconsistent workflow status caused by upload payload exception d…

### PR DESCRIPTION
**Pull request by jxu-nflx**
…uring workflow termination

Pull Request type
----

- [x ] Bugfix
- [ ] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe):

Changes in this PR
----
Catch exception of external storage upload failures during workflow termination and continue the execution to persist workflow status, etc.
_Describe the new behavior from this PR, and why it's needed_
Issue #2343 https://github.com/Netflix/conductor/issues/2323

Alternatives considered
----

_Describe alternative implementation you have considered_

### CODE REVIEW

File: core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java [hunk 1]
```diff
@@ -194,8 +194,16 @@ void failTask(Task task, PayloadType payloadType, String errorMsg) {
         throw new TerminateWorkflowException(errorMsg, Workflow.WorkflowStatus.FAILED, task);
     }
 
-    private void failWorkflow(String errorMsg) {
+    @VisibleForTesting
+    void failWorkflow(Workflow workflow, PayloadType payloadType, String errorMsg) {
         LOGGER.error(errorMsg);
-        throw new TerminateWorkflowException(errorMsg);
+        workflow.setReasonForIncompletion(errorMsg);
```
Comment (aravindanr) [inline core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java:200]: Lines L200- L206 does not look necessary as the catch block in [WorkflowExecutor.decide](https://github.com/Netflix/conductor/blob/main/core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java#L1162-L1165) and [WorkflowExecutor.terminateWorkflow](https://github.com/Netflix/conductor/blob/main/core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java#L849) set the status and reasonForIncompletion

File: core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java [hunk 2]
```diff
@@ -194,8 +194,16 @@ void failTask(Task task, PayloadType payloadType, String errorMsg) {
...
+        workflow.setStatus(Workflow.WorkflowStatus.FAILED);
+        if (payloadType == PayloadType.TASK_INPUT) {
```
Comment (aravindanr) [inline core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java:202]: From where the `failWorkflow` is called, the `PayloadType` should always be `WORKFLOW_INPUT` or `WORKFLOW_OUTPUT`.

**Review by aravindanr — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java [hunk 3]
```diff
@@ -837,7 +837,14 @@ public Workflow terminateWorkflow(Workflow workflow, String reason, String failu
             if (workflow.getWorkflowDefinition() == null) {
                 workflow = metadataMapperService.populateWorkflowWithDefinitions(workflow);
             }
-            deciderService.updateWorkflowOutput(workflow, null);
+
+            try {
+                deciderService.updateWorkflowOutput(workflow, null);
+            } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (jxu-nflx) [reply core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java:843]: The `verifyAndUpload` method could also throw `ApplicationException`: https://github.com/Netflix/conductor/blob/main/core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java#L87 So I use the generic catch of all exceptions here as I feel no matter what exception we got at this step, it shouldn't prevent us from continuing the termination execution

File: core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java (hunk 1 above)
Comment (jxu-nflx) [reply core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java:200]: I was going to remove that. But then on a second thought, I think if we get any exception from uploading outputdata to external storage during the terminate step, the final status should be `failed` instead of `terminated` with specific error message reflects that. What do you think?

File: core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java (hunk 2 above)
Comment (jxu-nflx) [reply core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java:202]: yeah the `payloadType` is passed into the `verifyAndUpload` method: https://github.com/Netflix/conductor/blob/main/core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java#L87 And we always verify both input and output data: https://github.com/Netflix/conductor/blob/main/core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java#L566

File: core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java (hunk 1 above)
Comment (aravindanr) [reply core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java:200]: I think the final status should be `TERMINATED`. A workflow is marked as FAILED, because of a Task failure or it timing out, not because of internal errors during the workflow evaluation. The `reasonForCompletion` should also be the original error that triggered the termination, not the internal error.

File: core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java (hunk 2 above)
Comment (aravindanr) [reply core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java:202]: Yes. `ExternalPayloadStorageUtils.verifyAndUpload()` is never called with a `Workflow` object and `PayloadType` as `TASK_INPUT` or `TASK_OUTPUT`. Workflow's input/output should not be manipulated here since the caller might be already manipulating it.

For instance, if the workflow definition has a `failure_workflow` defined, the workflow output is manipulated and setting it to `null` here would cause a `NullPointerException` there.

Long story short, i don't see a reason to update the `failWorkflow` method to fix this issue.

**Review by aravindanr — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java (hunk 2 above)
Comment (jxu-nflx) [reply core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java:202]: I added the logic to setting workflow output or input to `null` in the `failWorkflow` method because in normal circustances (i.e. when input or output size don't exceed threshold), we always do set the input/output to `null`: https://github.com/Netflix/conductor/blob/main/core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java#L162 So to be consistent, I think it makes sense to have the same logic added for `failWorkflow`: https://github.com/Netflix/conductor/blob/main/core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java#L135

File: core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java (hunk 1 above)
Comment (jxu-nflx) [reply core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java]: Sure ya updated
